odoo/addons/base/models/res_users.py

```diff
@@ -1178,6 +1216,15 @@ def _add_reified_groups(self, fields, values):
             elif is_selection_groups(f):
                 selected = [gid for gid in get_selection_groups(f) if gid in gids]
                 values[f] = selected and selected[-1] or False
+                #When the user is in at least two groups of category == 'User types' (Internal User or Portal or Public)
+                #The selected value must be the group with the most access, ie : Internal User > Portal > Public
+                #This is made to show the correct user type
+                if len(selected) > 1 :
+                    selectedGrps = self.env['res.groups'].browse(selected)
+                    if selectedGrps[0].category_id.xml_id == 'base.module_category_user_type':
+                        valInternal = next((grp.id for grp in selectedGrps if grp == self.env.ref('base.group_user',raise_if_not_found=False)), False)
```

Comment:
Maybe this is a bit more Odoo-style than creating a list of one element and taking the first one:
`selectedGrps.filtered(lambda g: g == self.env.ref('base.group_user', raise_if_not_found=False))[:1].id`